tests/test-dladdr.c

```diff
@@ -0,0 +1,163 @@
+    }
+    else
+    {
+        int sym_match  = info.dli_sname && strcmp( addrsym, info.dli_sname ) == 0;
+        int addr_match  = addr == info.dli_saddr;
+        passed = (expected_result == Pass && sym_match && addr_match)
+                 || (expected_result == PassWithoutSymbol && addr_match)
+                 || (expected_result == Fail && !(sym_match && addr_match));
+        printf( "check address %p with has symbol '%s' -> %s\n",addr, addrsym, passed ? "passed" : "failed" );
+        if (verbose)
+            print_dl_info( &info );
+    }
+    return !result;
+}
+
+/**
+ * @brief return address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @return pointer to symbol address or NULL
+ */
+void *get_symbol_address( char *libname, char *sym )
+{
+    void *library = NULL;
+    void *addr = NULL;
+
+    library = dlopen( libname, RTLD_GLOBAL );
+    if ( library == NULL )
+    {
+        fprintf( stderr, "could not open '%s'\n", libname );
+        return NULL;
+    }
+
+    addr = dlsym( library, sym );
+    dlclose( library );
+    if (!addr)
+        fprintf( stderr, "could not get address for symbol '%s'\n", sym );
+    return addr;
+}
+
+/**
+ * @brief check address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @param should_match result should match the given values
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr_by_dlopen( char *libname, char *sym, int should_match )
+{
+    void *addr = get_symbol_address( libname, sym);
+    if (!addr)
+        return 1;
+    return check_dladdr( addr, sym, should_match );
+}
+
+#ifdef _WIN32
+#define HMODULE void*
+#define HANDLE void*
+#define LPCVOID void*
+#define LPCSTR char*
+#define DWORD long
+#define SIZE_T long
+
+#ifdef _MSC_VER
+#define STDCALL __stdcall
+#else
+#define STDCALL __attribute__((__stdcall__))
+#endif
+
+/* link to import thunk */
+HMODULE STDCALL GetModuleHandleA (LPCSTR lpModuleName);
+/* link to directly to iat */
+__declspec(dllimport) HMODULE STDCALL LoadLibraryExA (LPCSTR lpLibFileName, HANDLE hFile, DWORD dwFlags);
+SIZE_T STDCALL VirtualQuery (LPCVOID lpAddress, LPCVOID lpBuffer, SIZE_T dwLength);
+#define PassWithoutSymbolOnWin PassWithoutSymbol
+#else
+#define PassWithoutSymbolOnWin Pass
+#endif
+
+
+int main(int argc, char **argv)
+{
+    int  result = 0;
+    if (argc == 2)
+        verbose = 1;
+
+    /* null pointer */
+    result = check_dladdr((void*)0, "null" , NoInfo);
+    /* invalid pointer */
+    result |= check_dladdr((void*)0x125, "invalid pointer" , Fail);
+    /* -ldl */
+    result |= check_dladdr( ((void*)dladdr), "dladdr" , Pass );
+    /* -ldl */
+    result |= check_dladdr( (void*)dlopen, "dlopen", Pass );
+    /* -lglibc */
+    result |= check_dladdr( (void*)vsnprintf, "vsnprintf", PassWithoutSymbolOnWin );
```

Comment:
This `vsnprintf` test is problematic. On windows it _passes without symbol_ because windows msvcrt does not provide vsnprintf implementation. Mingw/gcc therefore emulate it via its own library which is statically linked into final executable.

Microsoft started providing `vsnprintf` implementation only in recent years and only as part of ucrt (not msvcrt). Therefore this test starts failing with ucrt.

I would suggest to change this test to `atoi` stdlib function which is provided by msvcrt and therefore it will pass on windows also with symbol name.

I/O functions are problematic for testing as in msvcrt they are either not present or are not C99/C11 compliant. Therefore recent gcc/mingw uses own implementation, redefined in include files which avoid usage of msvcrt.